### PR TITLE
Add permutation method for array

### PR DESCRIPTION
**Pull request by jalyna**
Adds `Array#permutation`

### CODE REVIEW

**Comment by asterite** [comment]
Thanks!!

A few comments:

1. I think the block version should yield the permutations but not return an array of them (just return `self`). Ruby behaves that way:

    ```
    irb(main):003:0> a = [1, 2, 3]
    => [1, 2, 3]
    irb(main):004:0> x = 0
    => 0
    irb(main):005:0> a.permutation { x += 1 }
    => [1, 2, 3]
    irb(main):006:0> x
    => 6
    ```
2. Be careful with `self - [e]`, because it deletes all occurrences of `e`. So for example if you try `[1, 1, 2].permutation.to_a` in Ruby you get `[[1, 1, 2], [1, 2, 1], [1, 1, 2], [1, 2, 1], [2, 1, 1], [2, 1, 1]]`, but if I try that with your code I get `[]`.
3. Ruby's does this faster. But don't worry about this now, we can always improve this later as long as the API is compatible :-)

**Comment by jalyna** [comment]
@asterite Thank you for your fast feedback. I will adjust the specs & the code when I'm back from work :)

**Comment by jalyna** [comment]
@asterite I now added specs for the bug and the block behaviour

File: spec/std/array_spec.cr [hunk 1]
```diff
@@ -1107,4 +1107,21 @@ describe "Array" do
...
+
+  describe "permutation" do
+    assert { [1, 2, 2].permutation.should eq([[1,2,2],[1,2,2],[2,1,2],[2,2,1],[2,1,2],[2,2,1]]) }
+    assert { [1, 2, 3].permutation.should eq([[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]) }
+    assert { [1, 2, 3].permutation(1).should eq([[1],[2],[3]]) }
+    assert { [1, 2, 3].permutation(2).should eq([[1,2],[1,3],[2,1],[2,3],[3,1],[3,2]]) }
+    assert { [1, 2, 3].permutation(3).should eq([[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]) }
+    assert { [1, 2, 3].permutation(0).should eq([[] of Int32]) }
+    assert { [1, 2, 3].permutation(4).should eq([] of Array(Int32)) }
+    it "accepts a block" do
```
Comment (PragTob) [inline spec/std/array_spec.cr]: I'd fancy a newline here (before the it but that's fairly minor)

File: src/array.cr [hunk 2]
```diff
@@ -669,6 +669,57 @@ class Array(T)
...
+  #     a = [1, 2, 3]
+  #     sums = [] of Int32
+  #     a.permutation(2) { |p| sums << p.sum } #=> [1, 2, 3]
+  #     sums #=> [3, 4, 3, 5, 4, 5]
+  #
+  def permutation(n = nil : Int)
+    n ||= count
+    raise ArgumentError.new("size must be positive") if n < 0
+    ary = [] of Array(T)
+    return [[] of T] if n == 0
```
Comment (PragTob) [inline src/array.cr]: these first four lines of the method are completely duplicated, also a lot of the implementation in the each_with_index block is. Not sure if that's okay as we don't want to introduce helper methods as they would get into the name space.

Best solution might be to make one version of this method call the other (e.g. maybe with an empty block or something).

@asterite ?

**Comment by PragTob** [comment]
Looks good to me overall, minor comment about code layout and then once comment about code duplication :)

**Comment by asterite** [comment]
Looks good to me too! :-)

I checked the performance against Ruby, it's a bit slower. This is such an interesting/fun problem that a couple of days ago I also decided to implement it but for the case of `n = length`, but didn't think of the general case. This is my (dummy) [solution](http://play.crystal-lang.org/#/r/ak2). Looking at your solution, @jalyna, it's basically the idea I had but with a twist (the one needed for the general case).

I tried to see what Ruby does. First, for the cases of `n = 0` and `n = 1` there's something [easier](https://github.com/ruby/ruby/blob/trunk/array.c#L4973-L4983) that can be done. So I did that and the times improved, but still slower than Ruby.

Then there's the general case, handled in [permute0](https://github.com/ruby/ruby/blob/trunk/array.c#L4848-L4893) in Ruby. I didn't have time to analyze that code, but I think it basically permutes things in-place and then generates arrays from that, so you effectively don't create intermediate arrays for the intermediate results.

Then I remembered that @kostya mentioned that [he implemented that for Topaz](https://github.com/manastech/crystal/issues/733#issuecomment-109009689). When I first saw that I saw `to_a.reverse` and I thought "Hm, there you have two intermediate arrays, so I guess this won't be a fast solution", so back then I didn't try it. But now I decided to try it. I copied the code to a file, removed the `self` and used it. This is [the result](http://play.crystal-lang.org/#/r/ak3).

I was so surprised that it worked without changes, it's always amazing when you can copy/paste Ruby code and it just works.

Then I re-wrote that as an Array method, added a version that returns an array (because @kostya's solution yields the results) and benchmarked it: computing all permutation of a 10-sized array takes 1.76s in Ruby and 0.51s in Crystal with @kostya's version. Super fast!

So, I guess the correct thing to do here would be to use @kostya's version, but I'd like to understand it first :-)

@jalyna I hope you agree with this decision. One thing I can do is merge your pull request and then apply the optimizations, so we reuse your specs and check that they still work. In any case, if we didn't do that, eventually @kostya would have sent a pull request with optimizations (by now I think he's known in this project (and maybe others) as the king of optimizations: https://github.com/kostya/benchmarks, https://github.com/kostya/crystal-benchmarks-game, and he always tries to optimize things more and more).

One issue still remains: in Ruby `Array#permutation` (without a block) returns an Enumerator. Maybe we'd like to do the same and return an [Iterator](http://crystal-lang.org/api/Iterator.html), but I don't know how easy it would be to translate the algorithm to an iterator one.

**Comment by jhass** [comment]
Another implementation to look at is [Rubinius'](https://github.com/rubinius/rubinius/blob/master/kernel/common/array.rb#L994)

**Comment by kirbyfan64** [comment]
Also, I'm pretty sure [PyPy's is worth checking out](https://bitbucket.org/pypy/pypy/src/a0444ba5fc6ef4c5c8430d0ad8e1d0e51afdd685/pypy/module/itertools/interp_itertools.py?at=default#interp_itertools.py-1311), after all, it *is* the fastest Python implementation.

**Comment by asterite** [comment]
And while we are at this, what do you think about not using Ruby's names and using:

* `permutations`: returns an array with the permutations
* `each_permutation(&block)`: yields each permutation
* `each_permutation`: (without a block) returns an iterator (if we ever implement this)

**Comment by kostya** [comment]
i compared such code with ruby, topaz and crystal (topaz imll):
```ruby
a = (0..(ARGV[0] || 10).to_i).to_a
n = (ARGV[1] || a.length).to_i
nn = n / 2

t = Time.now
c = 0
w = 0
a.permutation(n) { |r| c += 1; w += r[nn] }
puts c
puts w
puts Time.now - t
```

ruby perm.rb 11 8: `7.10s`
topaz perm.rb 11 8: `2.17s`
crystal build perm.cr --release && ./perm 11 8:  `4.28s`

**Comment by kirbyfan64** [comment]
Well, this is the PyPy algorithm, ported to Ruby:

```ruby
class Array
  def self.permutation2(array, r=nil)
    r = array.length if r.nil?
    n = array.length
    if n-r < 0
      # Do something here (an error?)
      return
    else
      indices = (0...n).to_a
      cycles = n.downto(n-r).to_a
    end
    nxt = false
    while true
      result = (0...r).map {|i| indices[i]}
      i = r-1
      while i>=0
        j = cycles[i]-1
        if j>0
          cycles[i] = j
          indices[i], indices[-j] = indices[-j], indices[i]
          yield result
          nxt = true
          break
        else
          cycles[i] = indices.length-i
          num = indices[i]
          n1 = indices.length-1
          (i...n1).each {|k| indices[k] = indices[k+1] }
          indices[n1] = num
          i -= 1
        end
      end
      if nxt
        nxt = false
      else
        yield result
        return
      end
    end
  end
end

Array.permutation2([1, 2, 3], 2) {|r| puts "#{r}" }
```

**Comment by asterite** [comment]
@kostya which of all the codes did you compare?

**Comment by kostya** [comment]
ruby default, topaz my impl, and crystal topaz impl

i think topaz faster because jit reduces bounds checks, and also may be by faster gc.

**Comment by asterite** [comment]
I finally decided to merge this, do the renames and apply @kostya 's optimizations.

Thanks @jalyna for the initial implementation, the specs and docs! You sure opened an interesting discussion :-)

And thanks @kostya for the optimized version. I tried installing topaz but failed, I get this:

```
$ py.test
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/_pytest/config.py", line 543, in importconftest
    mod = conftestpath.pyimport()
  File "/usr/local/lib/python2.7/site-packages/py/_path/local.py", line 650, in pyimport
    __import__(modname)
  File "/Users/asterite-manas/Projects/topaz/tests/conftest.py", line 3, in <module>
    from rpython.config.translationoption import get_combined_translation_config
ImportError: No module named rpython.config.translationoption
ERROR: could not load /Users/asterite-manas/Projects/topaz/tests/conftest.py
```

Maybe topaz is faster because of the GC. We could try to optimize it, but right now it's much faster than Ruby so that's good enough :-)

**Pull request merged**